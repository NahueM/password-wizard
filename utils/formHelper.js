
function hasOneCapital(password){
    return (/[A-Z]/.test(password));
}

function hasSpecialCharacter(password){
    return (/[^a-zA-Z\d]/.test(password))
}

export function checkIsValidPassword(errors){
   const checks = Object.values(errors).find(err => err.status === false)
    if(checks === undefined){
        return true
    }else{
        return false
    }
}

export function checkAreSamePassword (confirmErrMsg, form, setAreSamePassword) {
    const{password, confirmPassword} = form
    let error = {...confirmErrMsg}
    if(Object.is(password, confirmPassword)){
        setAreSamePassword(true)
        return error = {...error, samePassword:{...error.samePassword, status:true}}
    }
     setAreSamePassword(false)
    return error = {...error, samePassword:{...error.samePassword, status:false}}
}

export function checkPassword(errMsg, password, setIsInValidPassword){
    let errors = {...errMsg}
    if(password.length > 7){
        errors = {...errors, cantCharacter:{...errors.cantCharacter, status:true}}
    }else{
        errors = {...errors, cantCharacter:{...errors.cantCharacter, status:false}}
    }

    if(hasOneCapital(password)){
        errors = {...errors, uppercase:{...errors.uppercase, status:true}}
    }else{
        errors = {...errors, uppercase:{...errors.uppercase, status:false}}
    }

    if(hasSpecialCharacter(password)){
        errors = {...errors, symbol:{...errors.symbol, status:true}}
    }else{
        errors = {...errors, symbol:{...errors.symbol, status:false}}
    }

    setIsInValidPassword(checkIsValidPassword(errors))
    return errors
}