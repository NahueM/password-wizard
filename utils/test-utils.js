
import * as React from 'react'
import {render as rtlRender} from '@testing-library/react'
import {WizardContextProvider} from '../hooks/useWizardContext'


function render(ui, { ...options } = {}) {
    // eslint-disable-next-line react/prop-types
    const Wrapper = ({ children }) => (
        <WizardContextProvider>{children}</WizardContextProvider>
    );
    return rtlRender(ui, { wrapper: Wrapper, ...options });
}

export * from '@testing-library/react'

export {render}