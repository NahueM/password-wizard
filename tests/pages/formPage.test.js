/**
 * @jest-environment jsdom
 */
 import {render, screen, waitForElementToBeRemoved} from '../../utils/test-utils'
 import userEvent from '@testing-library/user-event';
 import FormPage from "../../pages/formPage";


test('Should match snapshot', () => {
    const { asFragment } = render(<FormPage />);
    expect(asFragment()).toMatchSnapshot();
});
test("password checks should be red and next button should be disabled", () => {
    render(<FormPage />);
    expect(screen.getByText(/min 8 characters/i)).toHaveClass('text-red-500')
    expect(screen.getByText(/at least 1 symbol/i)).toHaveClass('text-red-500')
    expect(screen.getByText(/at least 1 capital letter/i)).toHaveClass('text-red-500')
    expect(screen.getByRole('button', {  name: /next/i})).toHaveAttribute('disabled')
});

test("min 8 character should change to green if you type 8 characters", () => {
    render(<FormPage />);
    expect(screen.getByText(/min 8 characters/i)).toHaveClass('text-red-500')
    userEvent.type(screen.getByPlaceholderText(/new password/i), 'loremipss');
    expect(screen.getByText(/min 8 characters/i)).toHaveClass('text-green-500')
    expect(screen.getByRole('button', {  name: /next/i})).toHaveAttribute('disabled')
}); 

test("At least 1 symbols should change to green", () => {
    render(<FormPage />);
    expect(screen.getByText(/at least 1 symbol/i)).toHaveClass('text-red-500')
    userEvent.type(screen.getByPlaceholderText(/new password/i), '$');
    expect(screen.getByText(/at least 1 symbol/i)).toHaveClass('text-green-500')
    expect(screen.getByRole('button', {  name: /next/i})).toHaveAttribute('disabled')
}); 

test("At least 1 capital letter should change to green", () => {
    render(<FormPage />);
    expect(screen.getByText(/at least 1 capital letter/i)).toHaveClass('text-red-500')
    userEvent.type(screen.getByPlaceholderText(/new password/i), 'P');
    expect(screen.getByText(/at least 1 capital letter/i)).toHaveClass('text-green-500')
    expect(screen.getByRole('button', {  name: /next/i})).toHaveAttribute('disabled')
}); 

test("All checks should be green and button still disabled", () => {
    render(<FormPage />);
    expect(screen.getByText(/min 8 characters/i)).toHaveClass('text-red-500')
    expect(screen.getByText(/at least 1 symbol/i)).toHaveClass('text-red-500')
    expect(screen.getByText(/at least 1 capital letter/i)).toHaveClass('text-red-500')

    userEvent.type(screen.getByPlaceholderText(/new password/i), 'Nahuel!"');

    expect(screen.getByText(/min 8 characters/i)).toHaveClass('text-green-500')
    expect(screen.getByText(/at least 1 symbol/i)).toHaveClass('text-green-500')
    expect(screen.getByText(/at least 1 capital letter/i)).toHaveClass('text-green-500')
    expect(screen.getByRole('button', {  name: /next/i})).toHaveAttribute('disabled')
}); 

test("All checks should be green and button still disabled", () => {
    render(<FormPage />);
    expect(screen.getByText(/min 8 characters/i)).toHaveClass('text-red-500')
    expect(screen.getByText(/at least 1 symbol/i)).toHaveClass('text-red-500')
    expect(screen.getByText(/at least 1 capital letter/i)).toHaveClass('text-red-500')

    userEvent.type(screen.getByPlaceholderText(/new password/i), 'Nahuel!"');

    expect(screen.getByText(/min 8 characters/i)).toHaveClass('text-green-500')
    expect(screen.getByText(/at least 1 symbol/i)).toHaveClass('text-green-500')
    expect(screen.getByText(/at least 1 capital letter/i)).toHaveClass('text-green-500')
    expect(screen.getByRole('button', {  name: /next/i})).toHaveAttribute('disabled')
}); 

test("typing diferent password 'Same password' should stay red and button disabled", () => {
    render(<FormPage />);
    expect(screen.getByText(/same password/i)).toHaveClass('text-red-500')
    
    userEvent.type(screen.getByPlaceholderText(/new password/i), 'Nahuel!"');
    userEvent.type(screen.getByPlaceholderText(/confirm/i), 'Nahuel"');

    expect(screen.getByText(/same password/i)).toHaveClass('text-red-500')
    expect(screen.getByRole('button', {  name: /next/i})).toHaveAttribute('disabled')
}); 

test("typing same password 'Same password' should change to green and button should be enable", async () => {
    render(<FormPage />);
    expect(screen.getByText(/same password/i)).toHaveClass('text-red-500')
    
    userEvent.type(screen.getByPlaceholderText(/new password/i), 'Nahuel!"');
    userEvent.type(screen.getByPlaceholderText(/confirm/i), 'Nahuel!"');

    expect(screen.getByText(/same password/i)).toHaveClass('text-green-500')
    expect(screen.getByRole('button', {  name: /next/i})).not.toHaveAttribute('disabled')
    userEvent.click(screen.getByRole('button', {  name: /next/i}))
}); 
