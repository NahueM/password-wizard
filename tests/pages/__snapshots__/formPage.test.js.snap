// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="text-secondary"
  >
    <header
      class="sticky top-0 z-50  bg-white  shadow-md p-5 md:px-10"
    >
      <div
        class="max-w-6xl mx-auto px-8 grid grid-cols-2"
      >
        <div
          class="relative flex items-center h-10 cursor-pointer my-auto pb-4"
        >
          <div
            style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
          >
            <noscript />
            <img
              alt="logo"
              decoding="async"
              href="/"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain; object-position: left;"
            />
          </div>
        </div>
        <div
          class="flex items-center justify-end"
        >
          <h1
            class="text-lg text-secondary md:text-3xl inline"
          >
            Password Wizard
          </h1>
        </div>
      </div>
    </header>
    <main
      class="max-w-6xl mx-auto px-8 sm:px-16"
    >
      <section
        class="pt-8"
      >
        <h1
          class="text-primary text-2xl font-semibold mb-8"
        >
          Create New Password
        </h1>
        <form
          class="flex flex-col p-5 border border-secondaryGrey"
        >
          <div
            class="flex flex-col lg:flex-row justify-between content-center pb-4"
          >
            <div
              class="flex flex-col space-y-2 w-full mb-4 lg:w-5/12"
            >
              <label>
                New Password
              </label>
              <div
                class="flex border-secondary items-center border "
              >
                <input
                  class=" outline-none p-2 w-full"
                  id="password"
                  placeholder="New Password"
                  type="password"
                  value=""
                />
                <svg
                  class="h-5 mr-4"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M3.707 2.293a1 1 0 00-1.414 1.414l14 14a1 1 0 001.414-1.414l-1.473-1.473A10.014 10.014 0 0019.542 10C18.268 5.943 14.478 3 10 3a9.958 9.958 0 00-4.512 1.074l-1.78-1.781zm4.261 4.26l1.514 1.515a2.003 2.003 0 012.45 2.45l1.514 1.514a4 4 0 00-5.478-5.478z"
                    fill-rule="evenodd"
                  />
                  <path
                    d="M12.454 16.697L9.75 13.992a4 4 0 01-3.742-3.741L2.335 6.578A9.98 9.98 0 00.458 10c1.274 4.057 5.065 7 9.542 7 .847 0 1.669-.105 2.454-.303z"
                  />
                </svg>
              </div>
              <div
                class="space-y-2 bg-gray-200 p-3"
                data-testid="password"
              >
                <div
                  class="text-xs flex space-x-4 text-red-500"
                >
                  Min 8 characters
                </div>
                <div
                  class="text-xs flex space-x-4 text-red-500"
                >
                  At least 1 symbol
                </div>
                <div
                  class="text-xs flex space-x-4 text-red-500"
                >
                  At least 1 capital letter
                </div>
              </div>
            </div>
            <div
              class="flex flex-col space-y-2 w-full undefined lg:w-5/12"
            >
              <label>
                Confirm your password
              </label>
              <div
                class="flex border-secondary items-center border "
              >
                <input
                  class=" outline-none p-2 w-full"
                  id="confirmPassword"
                  placeholder="Confirm"
                  type="password"
                  value=""
                />
                <svg
                  class="h-5 mr-4"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M3.707 2.293a1 1 0 00-1.414 1.414l14 14a1 1 0 001.414-1.414l-1.473-1.473A10.014 10.014 0 0019.542 10C18.268 5.943 14.478 3 10 3a9.958 9.958 0 00-4.512 1.074l-1.78-1.781zm4.261 4.26l1.514 1.515a2.003 2.003 0 012.45 2.45l1.514 1.514a4 4 0 00-5.478-5.478z"
                    fill-rule="evenodd"
                  />
                  <path
                    d="M12.454 16.697L9.75 13.992a4 4 0 01-3.742-3.741L2.335 6.578A9.98 9.98 0 00.458 10c1.274 4.057 5.065 7 9.542 7 .847 0 1.669-.105 2.454-.303z"
                  />
                </svg>
              </div>
              <div
                class="space-y-2 bg-gray-200 p-3"
                data-testid="confirmPassword"
              >
                <div
                  class="text-xs flex space-x-4 text-red-500"
                >
                  Same password
                </div>
              </div>
            </div>
          </div>
          <div
            class="flex flex-col space-y-2 w-full undefined lg:w-5/12"
          >
            <label>
              Phrase Clue
            </label>
            <div
              class="flex border-secondary items-center border "
            >
              <input
                class=" outline-none p-2 w-full"
                id="phrase"
                placeholder=" Will help you remember password"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="flex flex-row w-full pt-5 space-x-4 sm:justify-between"
          >
            <button
              class="w-full py-2 px-4 border  border-transparent shadow-sm font-medium rounded-md text-white  bg-primary disabled:bg-gray-400 hover:bg-red-700 sm:w-40 md:w-52 text-lg "
              href="/"
              type="button"
            >
              Back
            </button>
            <button
              class="w-full py-2 px-4 border  border-transparent shadow-sm font-medium rounded-md text-white  bg-primary disabled:bg-gray-400 hover:bg-red-700 sm:w-40 md:w-52 text-lg "
              disabled=""
              type="button"
            >
              Next
            </button>
          </div>
        </form>
      </section>
    </main>
  </div>
</DocumentFragment>
`;
